src/AMSlib/wf/basedb.hpp

```diff
@@ -1538,48 +1459,70 @@ class RabbitMQDB final : public BaseDB<TypeValue>
           "dimensions (%d, %d)",
           type().c_str(),
           num_elements,
           inputs.size(),
           outputs.size())
 
-    const size_t inputs_size = num_elements * inputs.size();
-    auto inputs_data = flatten_features(num_elements, inputs, true);
-    DBG(RabbitMQDB,
-        "[store(%d, %d, %d)] input sent %d B",
-        num_elements,
-        inputs.size(),
-        outputs.size(),
-        inputs_size * sizeof(TypeValue))
-    _nb_msg_send++;
-    _evbuffer->push(static_cast<void*>(inputs_data),
-                    inputs_size * sizeof(TypeValue));
-
-    // TODO: investigate that
-    // Necessary for some reasons, other the event buffer overheat
-    // "[err] buffer.c:1066: Assertion chain || datlen==0 failed in
-    // evbuffer_copyout" potentially segfault, and with CUDA could also lead to
-    // packet losses
-    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
-
-    const size_t outputs_size = num_elements * outputs.size();
-    auto outputs_data = flatten_features(num_elements, outputs, true);
-    DBG(RabbitMQDB,
-        "[store(%d, %d, %d)] output sent %d B",
-        num_elements,
-        inputs.size(),
-        outputs.size(),
-        outputs_size * sizeof(TypeValue))
-    _nb_msg_send++;
-    _evbuffer->push(static_cast<void*>(outputs_data),
-                    outputs_size * sizeof(TypeValue));
-
-    free(inputs_data);
-    free(outputs_data);
+    const size_t num_in = inputs.size();
+    const size_t num_out = outputs.size();
+    const size_t header_size = 2*sizeof(uint8_t) + sizeof(uint32_t) + 3*sizeof(uint16_t);
+    const size_t blob_size = header_size + (num_elements * (num_in + num_out)) * sizeof(TypeValue);
+    uint8_t* data_blob = static_cast<uint8_t*>(std::malloc(blob_size*sizeof(uint8_t)));
+
+    uint16_t mpirank = static_cast<uint16_t>(_rank);
+    uint32_t num_elem = static_cast<uint32_t>(num_elements);
+    uint16_t input_dim = static_cast<uint16_t>(num_in);
+    uint16_t output_dim = static_cast<uint16_t>(num_out);
+
+    size_t current_offset = 0;
+    data_blob[current_offset] = static_cast<uint8_t>(header_size); // Header size (should be 12 bytes here)
+    current_offset += sizeof(uint8_t);
+
+    data_blob[current_offset] = static_cast<uint8_t>(sizeof(TypeValue)); // Data type (should be 1 bytes)
+    current_offset += sizeof(uint8_t);
+
+    std::memcpy(data_blob+current_offset, &(mpirank), sizeof(uint16_t)); // MPI rank (should be 2 bytes)
+    current_offset += sizeof(uint16_t);
+
+    std::memcpy(data_blob+current_offset, &(num_elem), sizeof(uint32_t)); // Num elem (should be 4 bytes)
+    current_offset += sizeof(uint32_t);
+
+    std::memcpy(data_blob+current_offset, &(input_dim), sizeof(uint16_t)); // Input dim (should be 2 bytes)
+    current_offset += sizeof(uint16_t);
+
+    std::memcpy(data_blob+current_offset, &(output_dim), sizeof(uint16_t)); // Output dim (should be 2 bytes)
+    current_offset += sizeof(uint16_t);
```

Comment:
Create a function that does this (create_hear_in_blob) or something.

Reply:
Good point.